src/main/java/org/opensearch/ad/ml/CheckpointDao.java

```diff
@@ -455,26 +568,70 @@ public void restoreModelCheckpoint(String modelId, ActionListener<Optional<Entry
     }
 
     /**
-     * Returns to listener the checkpoint for the model.
+     * Returns to listener the checkpoint for the rcf model.
      *
      * @param modelId id of the model
      * @param listener onResponse is called with the model checkpoint, or empty for no such model
      */
-    public void getModelCheckpoint(String modelId, ActionListener<Optional<String>> listener) {
-        clientUtil
-            .<GetRequest, GetResponse>asyncRequest(
-                new GetRequest(indexName, modelId),
-                client::get,
-                ActionListener.wrap(response -> listener.onResponse(processModelCheckpoint(response)), listener::onFailure)
-            );
+    public void getRCFModel(String modelId, ActionListener<Optional<RandomCutForest>> listener) {
+        clientUtil.<GetRequest, GetResponse>asyncRequest(new GetRequest(indexName, modelId), client::get, ActionListener.wrap(response -> {
+            Optional<Object> rcfCheckpoint = processRCFModelCheckpoint(response);
+            if (!rcfCheckpoint.isPresent()) {
+                listener.onFailure(new ResourceNotFoundException("", "Fail to find model " + modelId));
+                return;
+            }
+            listener
+                .onResponse(
+                    rcfCheckpoint
+                        .map(
+                            checkpoint -> AccessController
+                                .doPrivileged((PrivilegedAction<RandomCutForest>) () -> restoreRCFModel((String) checkpoint))
+                        )
+                );
+        }, listener::onFailure));
+    }
+
+    /**
+     * Returns to listener the checkpoint for the threshold model.
+     *
+     * @param modelId id of the model
+     * @param listener onResponse is called with the model checkpoint, or empty for no such model
+     */
+    public void getThresholdModel(String modelId, ActionListener<Optional<ThresholdingModel>> listener) {
```

Comment:
Question: If the threshold model was also updated to V2 (say?) would this suffice? Any code we want to leave here such that V2 of Threshold model adheres to that?

Reply:
Great question.  Currently, for single stream detectors, theshold model uses model (old) field in checkpoint; for HCAD detectors, threshold models use modelV2 (new)  (as for HCAD rcf, threshold, samples are in the same checkpoint.  Since rcf changed to modelV2, threshold uses modelV2 as well.).  When there is a new threshold model, there can be a few ways to migrate: for example, if the new threshold model only stores 5 numbers, we can store them in 5 different fields and fallback to model/modelV2 field if these 5 fields are not found.  Or we introduce a modelV3 field for the new model format.